sw/device/lib/dif/dif_i2c.h

```diff
@@ -0,0 +1,390 @@
+   * larger if desired.
+   *
+   * Setting this value to zero will result in the minimum period being used.
+   */
+  uint32_t scl_period_nanos;
+} dif_i2c_timing_config_t;
+
+/**
+ * Represents an I2C device. This struct should only be initialized using
+ * `dif_i2c_init()`.
+ */
+typedef struct dif_i2c { mmio_region_t base_addr; } dif_i2c_t;
+
+typedef enum dif_i2c_result {
+  kDifI2cResultOk = 0,
+  kDifI2cResultBadArg,
+} dif_i2c_result_t;
+
+/**
+ * Represents a valid watermark level for one of the I2C FIFOs.
+ */
+typedef enum dif_i2c_watermark_level {
+  kDifI2cLevel1Byte = 0,
+  kDifI2cLevel4Byte,
+  kDifI2cLevel8Byte,
+  kDifI2cLevel16Byte,
+  kDifI2cLevel30Byte,
+} dif_i2c_level_t;
+
+/**
+ * Represents an I2C-related interrupt type.
+ */
+typedef enum dif_i2c_irq_type {
+  /**
+   * Fired when the FMT FIFO underflows its watermark.
+   */
+  kDifI2cIrqTypeFmtWatermarkUnderflow = 0,
+  /**
+   * Fired when the RX FIFO overflows its watermark.
+   */
+  kDifI2cIrqTypeRxWatermarkOverflow,
+  /**
+   * Fired when the FMT FIFO overflows.
+   */
+  kDifI2cIrqTypeFmtFifoOverflow,
+  /**
+   * Fired when the RX FIFO overflows.
+   */
+  kDifI2cIrqTypeRxFifoOverflow,
+  /**
+   * Fired when there is no ACK in response to an address or data write.
+   */
+  kDifI2cIrqTypeNak,
+  /**
+   * Fired when the SCL line seems to have interference.
+   */
+  kDifI2cIrqTypeSclInterference,
+  /**
+   * Fired when the SDA line seems to have interference.
+   */
+  kDifI2cIrqTypeSdaInterference,
+  /**
+   * Fired when the target streches the clock beyond the allowed period.
+   */
+  kDifI2cIrqTypeClockStretchTimeout,
+  /**
+   * Fired when the target does not maintain a stable SDA line.
+   */
+  kDifI2cIrqTypeSdaUnstable,
+} dif_i2c_irq_type_t;
+
+/**
+ * Represents an enablement state.
+ */
+typedef enum dif_i2c_enable {
+  kDifI2cEnabled,
+  kDifI2cDisabled,
+} dif_i2c_enable_t;
+
+/**
+ * Computes timing parameters for an I2C device using the input parameters in
+ * `config`.
+ *
+ * @param config Configuration values for producing timing parameters.
+ * @param out Out-param for the actual timing parameters.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_compute_timing(dif_i2c_timing_config_t config,
+                                        dif_i2c_timing_params_t *out);
+
+/**
+ * Initializes an I2C device using the given timing parameters.
+ *
+ * @oaram base_addr The start of the I2C device register.
+ * @param timing Timing parameters to initialize with.
+ * @param i2c Out param for the initialized device.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_init(mmio_region_t base_addr,
+                              dif_i2c_timing_params_t timing, dif_i2c_t *i2c);
+
+/**
+ * Resets the state of the RX FIFO, essentially dropping all recieved bytes.
+ *
+ * @param i2c An I2C devce.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_reset_rx_fifo(const dif_i2c_t *i2c);
```

Comment:
What are the use-cases for this function apart from `init`? Should it be an internal function instead? Same for the `fmt_fifo`.